File: homework3/Satyam_Mishra/Q5_FirstKBinaryNumbers.java
// DS/Algo used - BFS

// Time Complexity = O(k)
// Space Complexity = O(k)

package homework3.Satyam_Mishra;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

public class Q5_FirstKBinaryNumbers {
    private static List<String> bfs(int k) {
        List<String> res = new ArrayList<>();
        if(k == 0) return res;

        res.add("0");
        if(k == 1) return res;

        Queue<String> q = new ArrayDeque<>();
        q.add("1");
        res.add("1");

        while(res.size() < k) {
            String front = q.poll();
            
            String addZero = front + "0";
            String addOne = front + "1";

            q.add(addZero);
            res.add(addZero);

            if(res.size() < k) {
                q.add(addOne);
                res.add(addOne);
            }
        }

        return res;
    }

    public static void main(String[] args) {
        int k1 = 5;
        List<String> res1 = bfs(k1);
        System.out.println("First k1 binary numbers are: " + res1);

        int k2 = 10;
        List<String> res2 = bfs(k2);
        System.out.println("First k2 binary numbers are: " + res2);
    }
}

// time taken 20m 15s

File: homework3/Aditya_Prasad/q5_ FirstKBinaryNumbers.cpp
// DS/Algo used - BFS

// Time Complexity - O(k)

// Memory Complexity - O(k)

#include <bits/stdc++.h>
using namespace std;

void FirstKBinaryNumbers(int k){
    cout << "0" << " ";

    if (k == 1) return;

    queue<string> q;
    q.push("1");

    for (int i = 0; i < k - 1; i++){
        string curr = q.front();
        q.pop();

        cout << curr << " ";

        q.push(curr + "0");
        q.push(curr + "1");
    }
}

int main(){
    FirstKBinaryNumbers(5);
    FirstKBinaryNumbers(10);

}

// Time Taken - 20 min